### PR TITLE
Fully resolve ipfs paths down to its final hash

### PR DESCRIPTION
**Pull request by travisperson**
Given an ipfs path, this will resolve it down to it's furthest point.

Tests to come.

### CODE REVIEW

File: core/commands/resolve.go [hunk 1]
```diff
@@ -84,6 +84,23 @@ Resolve the value of another name recursively:
 			return
 		}
 
+		if len(output.Segments()) > 2 {
+			merkNode, err := n.Resolver.ResolvePath(n.Context(), output)
```
Comment (whyrusleeping) [inline core/commands/resolve.go]: i'm gonna use `merkNode` from now on.

**Comment by whyrusleeping** [comment]
what case wasnt being handled previously?

**Comment by travisperson** [comment]
Given a path after a hash it would just return the same input and not actually do anything.
Before
```
$ ipfs resolve /ipfs/Qmf2yWpPKY19GqqfCidT9cCKMZnpvGfo1tHdVAYHGs6b2z/something
/ipfs/Qmf2yWpPKY19GqqfCidT9cCKMZnpvGfo1tHdVAYHGs6b2z/something⏎
```
After
```
$ ipfs resolve /ipfs/Qmf2yWpPKY19GqqfCidT9cCKMZnpvGfo1tHdVAYHGs6b2z/something
/ipfs//ipfs/QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn⏎
```

**Comment by whyrusleeping** [comment]
cc @wking

**Comment by whyrusleeping** [comment]
also the output `/ipfs//ipfs/QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn` seems wrong. got an extra ipfs on there

**Comment by wking** [comment]
I want this feature (see also #1082), but I'm not currently clear on
the intended UI (see #1307).  And I also think we want to flip the
logic around and add an IPFS resolver (the stuff that's currently in
core/pathresolver.go) to our multi-protocol resolver (currently in
namesys/namesys.go) and teach the multi-protocol logic (currently in
namesys/base.go) to distinguish between symlinks (which replace the
looked-up path in the segment stack) and non-symlinks (which reduce
the segment stack).  See also [1].

Because as far as path-resolution goes, it doesn't matter if we're
using Merkle nodes, DHT links, DNS links, … to figure out what a path
means.

[1]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30386896

**Comment by travisperson** [comment]
> also the output /ipfs//ipfs/QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn seems wrong. got an extra ipfs on there

Copy and paste error, it it's not there when ran.

**Comment by whyrusleeping** [comment]
I like the functionality that this command has in this PR.

**Comment by whyrusleeping** [comment]
i must be really out of the loop. Didnt this command used to be `ipfs name resolve` ?

**Comment by jbenet** [comment]
@travisperson i think full resolution should be under a `-r / --recursive` flag.

```go
```

**Comment by jbenet** [comment]
@wking 

i think there's many parts here:
1. resolving object links (just ipfs objects)
2. resolving key names (via routing records)
3. resolving dns names (via dns)
4. resolving .bit names (via namecoin)
5. resolving onion names (via tor)
6. resolving all _names_ (everything currently under `/ipns/...`, key names and dns names, etc)
7. all of the above

today, we have:
-  (2., 3., and 6.) under `ipfs name resolve`, and that's fine.

**Comment by jbenet** [comment]
(btw, `ipfs object` will be renamed to `ipfs dag`

**Comment by travisperson** [comment]
> @travisperson i think full resolution should be under a -r / --recursive flag. (as @wking points out)

Prior to this PR, the resolve command did nothing for ipfs paths, it just returned the same input. As it is, it's not really a recursive resolve, unless you only want it to resolve a single folder at a time, such as:
```
$ ipfs resolve /ipfs/<hash>/path/to/end
/ipfs/<hash>/to/end
```

**Comment by wking** [comment]
On Tue, Jun 02, 2015 at 03:44:02PM -0700, Jeromy Johnson wrote:
> i must be really out of the loop. Didnt this command used to be
> `ipfs name resolve` ?

It did until #1208, and it likely will have that name again soon.
That's what #1307 is about.

**Comment by wking** [comment]
On Wed, Jun 03, 2015 at 12:12:25AM -0700, Juan Batiz-Benet wrote:
> i think there's many parts here:

I responded in #1307 to keep the intended interface conversation in
one place.

**Comment by wking** [comment]
On Wed, Jun 03, 2015 at 12:30:33AM -0700, Juan Batiz-Benet wrote:
> (btw, `ipfs object` will be renamed to `ipfs dag`

Spun off into #1328.

**Comment by wking** [comment]
On Wed, Jun 03, 2015 at 12:54:25AM -0700, Travis Person wrote:
> > i think full resolution should be under a -r / --recursive
> > flag. (as @wking points out)
> 
> Prior to this PR, the resolve command did nothing for ipfs paths, it
> just returned the same input. As it is, it's not really a recursive
> resolve, unless you only want it to resolve a single folder at a
> time, such as:
>
> $ ipfs resolve /ipfs/<hash>/path/to/end
> /ipfs/<hash>/to/end

This is how I think a non-recursive Merkle-link lookup should work, so
I want that in the all-protocol resolver.  Although I'd phrase it
like this:

  With the Merkle DAG:

    Object        Link
               Name Hash
    ========   ==== ========
    <hash-1>   path <hash-2>
    <hash-2>   to   <hash-3>
    <hash-3>   end  <hash-4>

  Then:

    $ ipfs resolve /ipfs/<hash-1>/path/to/end
    /ipfs/<hash-2>/to/end
    $ ipfs resolve --recursive /ipfs/<hash-1>/path/to/end
    /ipfs/<hash-4>

But I'd also like to see support for symlink objects with arbitrary
data (including any supported protocol):

  /ipns/<pk-hash>/child

Then we could have:

  With the Merkle DAG:

    Object     Type           Link        Data
                           Name Hash
    ========   ==========  ==== ========  =====
    <hash-1>   unixfs-dir  a    <hash-2>
    <hash-2>   unixfs-dir  b    <hash-3>
    <hash-3>   symlink                    /ipns/<pk-hash>/d
    <hash-4>   unixfs-dir  c    <hash-5>
    <hash-5>   fanout      d*   <hash-6>
    <hash-6>   fanout-leaf d    <hash-7>
    <hash-7>   unixfs-dir  e    <hash-8>


  And the DHT-namesys entries:

     Pubkey     Record
     =========  ==============
     <pk-hash>  /ipfs/<hash-4>/c

  Then:

    $ ipfs resolve /ipfs/<hash-2>/b/e
    /ipfs/<hash-3>/e
    $ ipfs resolve /ipfs/<hash-3>/e
    /ipns/<pk-hash>/d/e
    $ ipfs resolve /ipns/<pk-hash>/d/e
    /ipfs/<hash-4>/c/d/e
    $ ipfs resolve /ipns/<hash-4>/c/d/e
    /ipfs/<hash-5>/d/e
    $ ipfs resolve /ipns/<hash-5>/d/e
    /ipfs/<hash-7>/e
    $ ipfs resolve --recursive /ipfs/<hash-2>/b/e
    /ipfs/<hash-8>

Besides the link through IPNS space, this also shows the non-recursive
form silently traversing the <hash-5> fanout tree.

**Comment by jbenet** [comment]
> This is how I think a non-recursive Merkle-link lookup should work, so
> I want that in the all-protocol resolver.  Although I'd phrase it
> like this:

yes, i like that resolve sequence.

> But I'd also like to see support for symlink objects with arbitrary
> data (including any supported protocol):

Yes, that looks right to me. the links are a bit tricky. we need to think carefully how to "symlink" from a merkledag object to an ipns mutable record safely. The nice replicated dag guarantees are a bit broken if we dont do it carefully.

btw, dm (mazieres) pointed out to me that getting symlinks right was going to be hard. :) 

> this also shows the non-recursive
> form silently traversing the <hash-5> fanout tree.

yes agreed with this. this is actually similar to how git traverses `commit -> first tree` object. we need a good, reliable way to do this. btw, the fanout links thing should work like a union or a unixfs file-- each node should work as a valid root (i.e. no `fanout` and `fanout-leaf` distinction).

**Comment by jbenet** [comment]
@travisperson clear on next things to do?

**Comment by wking** [comment]
On Wed, Jun 03, 2015 at 02:32:47PM -0700, Juan Batiz-Benet wrote:
> > But I'd also like to see support for symlink objects with
> > arbitrary data (including any supported protocol):
>
> Yes, that looks right to me. the links are a bit tricky.

Symlinks are #1104.  Their impact here is just: if we want links out
of the immutable Merkle DAG at all, we'll need to figure out how
they'll be represented in the resolver UI.  For me, they motivate
moving Merkle-DAG resolution into NameSys [1].

> btw, the fanout links thing should work like a union or a unixfs
> file-- each node should work as a valid root (i.e. no `fanout` and
> `fanout-leaf` distinction).

Spun off into ipfs/specs#10.

[1]: https://github.com/ipfs/go-ipfs/pull/1320#issuecomment-108027794

**Comment by rht** [comment]
@whyrusleeping can be closed

**Comment by rht** [comment]
(here goes)

**Pull request closed without merging**